### DISCUSSION TITLE
Integration with Symfony - memory leak

### DISCUSSION THREAD

**Discussion by rmikalkenas** [roadrunner-server/roadrunner#1458] [Q&A] [answered]
Hi all!

Recently I started playing around RoadRunner and so far I'm impressed with all the capabilities it provides:)
While working with vanilla php examples everything works fine and the memory consumption is stable across multiple http requests.
But lately tried to run fresh Symfony installation and faced increasing memory between http requests. At first I thought it's because of custom php extensions I have locally (Xdebug and Datadog), but removing them didn't help. Afterwards I tried to execute `gc_collect_cycles();` after each request - the memory became stable, but overall performance degraded.

Curious is it expected behavior using Symfony framework and is `gc_collect_cycles();` the best approach for solving memory leaks..?
Maybe someone had similar issues and could guide me how to properly configure RR to use with Symfony?

Reproduced for memory leak: https://github.com/rmikalkenas/roadrunner-leak-reproducer
My local env:
```
rr -v               
rr version 2.12.2 (build time: 2023-01-12T12:07:45+0000, go1.19.5), OS: linux, arch: amd64

php -v
PHP 8.1.14 (cli) (built: Jan 11 2023 08:10:46) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.14, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.14, Copyright (c), by Zend Technologies
```

**Comment by rustatian** [comment, marked as answer]
Hey, @rmikalkenas 👋🏻 Nice to meet you, thanks for using RoadRunner ❤️ 

Have you tried `./rr workers -i` ? How did you measure memory consumption?

**Reply by rmikalkenas** [reply]
Hi @rustatian , I measure memory usage by returning `memory_get_usage();` value as response content: https://github.com/rmikalkenas/roadrunner-leak-reproducer/blob/master/src/Controller/HealthController.php#L15
And then executing http requests by manually running a php script: https://github.com/rmikalkenas/roadrunner-leak-reproducer/blob/master/debug.php

**Reply by rustatian** [reply]
I guess it's better to ask the author of this integration: https://github.com/Baldinof/roadrunner-bundle. You may also uncomment this section: https://github.com/rmikalkenas/roadrunner-leak-reproducer/blob/master/.rr.yaml#L34 and double check the memory consumption via `./rr workers -i` command. But also, I don't think that `gc_collect_cycles();` is expected behavior.

**Reply by rmikalkenas** [reply]
Some interesting findings:
`rr workers -i` indicates a stable memory consuption of 42MB (even on more than 50k+ execs) - so it's great!
But `memory_get_usage();` shows fluctuating memory usage:
I did 10k requests, after each `memory_get_usage();` increases, but once it reaches ~14MB it drops to ~5MB.. Looks like somebody is clearing the memory. To minimize impact of `Baldinof/roadrunner-bundle` I removed it completely and manually implemented worker based on [docs example](https://roadrunner.dev/docs/php-worker/2.x/en). The result was exactly the same. So the only assumption that comes to my mind is that php has some kind of additional GC clearing policy which is based on memory consumption.

![image](https://user-images.githubusercontent.com/14221532/218430292-e9d68593-9382-4ebe-8829-8e4ac3c11f76.png)

**Reply by rustatian** [reply]
Basically, `./rr workers -i` works like a supervisor and shows the correct RSS memory usage (because it is obtained from your OS). I'm not a PHP, so, that's hard for me to say, what's happening inside the PHP and what method to use, but I guess this is the PHP GC. It might be you need to check the `memory_limit` in your `php.ini`.

**Reply by wolfy-j** [reply]
memory_get_usage() shows memory allocated by PHP and it's affected by GC. So it's OK for it to drop, use RR memory metrics (see above) for a RSS usage which includes engine memory. That's more reliable way.

**Reply by rmikalkenas** [reply]
Appreciate for the help! Marked discussion as answered
